File: js/test/output-blocks.test.js
/**
 * Tests for output-blocks module
 *
 * Tests the "status spine" format: width-independent, lossless output
 */

const { describe, it, expect } = require('bun:test');

const {
  // Spine format exports
  SPINE,
  SUCCESS_MARKER,
  FAILURE_MARKER,
  createSpineLine,
  createEmptySpineLine,
  createCommandLine,
  getResultMarker,
  parseIsolationMetadata,
  generateIsolationLines,

  // Main block functions
  createStartBlock,
  createFinishBlock,
  formatDuration,

  // Links notation utilities
  escapeForLinksNotation,
  formatAsNestedLinksNotation,
} = require('../src/lib/output-blocks');

describe('output-blocks module', () => {
  describe('spine format constants', () => {
    it('should export spine character', () => {
      expect(SPINE).toBe('│');
    });

    it('should export result markers', () => {
      expect(SUCCESS_MARKER).toBe('✓');
      expect(FAILURE_MARKER).toBe('✗');
    });
  });

  describe('createSpineLine', () => {
    it('should create a line with spine prefix and padded label', () => {
      const line = createSpineLine('session', 'abc-123');
      expect(line).toBe('│ session   abc-123');
    });

    it('should pad labels to 10 characters', () => {
      const shortLabel = createSpineLine('exit', '0');
      expect(shortLabel).toBe('│ exit      0');

      const longLabel = createSpineLine('isolation', 'docker');
      expect(longLabel).toBe('│ isolation docker');
    });
  });

  describe('createEmptySpineLine', () => {
    it('should create just the spine character', () => {
      expect(createEmptySpineLine()).toBe('│');
    });
  });

  describe('createCommandLine', () => {
    it('should create a line with $ prefix', () => {
      expect(createCommandLine('echo hi')).toBe('$ echo hi');
    });
  });

  describe('getResultMarker', () => {
    it('should return success marker for exit code 0', () => {
      expect(getResultMarker(0)).toBe('✓');
    });

    it('should return failure marker for non-zero exit codes', () => {
      expect(getResultMarker(1)).toBe('✗');
      expect(getResultMarker(127)).toBe('✗');
      expect(getResultMarker(-1)).toBe('✗');
    });
  });

  describe('parseIsolationMetadata', () => {
    it('should parse environment and mode', () => {
      const extraLines = ['[Isolation] Environment: docker, Mode: attached'];
      const metadata = parseIsolationMetadata(extraLines);

      expect(metadata.isolation).toBe('docker');
      expect(metadata.mode).toBe('attached');
    });

    it('should parse session name', () => {
      const extraLines = ['[Isolation] Session: my-session'];
      const metadata = parseIsolationMetadata(extraLines);

      expect(metadata.session).toBe('my-session');
    });

    it('should parse docker image', () => {
      const extraLines = ['[Isolation] Image: ubuntu:latest'];
      const metadata = parseIsolationMetadata(extraLines);

      expect(metadata.image).toBe('ubuntu:latest');
    });

    it('should parse all fields together', () => {
      const extraLines = [
        '[Isolation] Environment: docker, Mode: detached',
        '[Isolation] Session: docker-container-123',
        '[Isolation] Image: node:18-alpine',
        '[Isolation] User: testuser (isolated)',
      ];
      const metadata = parseIsolationMetadata(extraLines);

      expect(metadata.isolation).toBe('docker');
      expect(metadata.mode).toBe('detached');
      expect(metadata.session).toBe('docker-container-123');
      expect(metadata.image).toBe('node:18-alpine');
      expect(metadata.user).toBe('testuser');
    });
  });

  describe('generateIsolationLines', () => {
    it('should generate lines for docker isolation', () => {
      const metadata = {
        isolation: 'docker',
        mode: 'attached',
        image: 'ubuntu',
        session: 'docker-container-1',
      };
      const lines = generateIsolationLines(metadata);

      expect(lines).toContain('│ isolation docker');
      expect(lines).toContain('│ mode      attached');
      expect(lines).toContain('│ image     ubuntu');
      expect(lines).toContain('│ container docker-container-1');
    });

    it('should generate lines for screen isolation', () => {
      const metadata = {
        isolation: 'screen',
        mode: 'attached',
        session: 'screen-session-1',
      };
      const lines = generateIsolationLines(metadata);

      expect(lines).toContain('│ isolation screen');
      expect(lines).toContain('│ mode      attached');
      expect(lines).toContain('│ screen    screen-session-1');
    });
  });

  describe('createStartBlock', () => {
    it('should create a start block with session and timestamp', () => {
      const block = createStartBlock({
        sessionId: 'test-uuid-1234',
        timestamp: '2025-01-01 00:00:00',
        command: 'echo hello',
      });

      expect(block).toContain('│ session   test-uuid-1234');
      expect(block).toContain('│ start     2025-01-01 00:00:00');
      expect(block).toContain('$ echo hello');
    });

    it('should include empty spine line before command', () => {
      const block = createStartBlock({
        sessionId: 'test-uuid',
        timestamp: '2025-01-01 00:00:00',
        command: 'echo hello',
      });

      const lines = block.split('\n');
      // Last line should be command, second-to-last should be empty spine
      expect(lines[lines.length - 1]).toBe('$ echo hello');
      expect(lines[lines.length - 2]).toBe('│');
    });

    it('should include isolation metadata when provided', () => {
      const block = createStartBlock({
        sessionId: 'test-uuid',
        timestamp: '2025-01-01 00:00:00',
        command: 'echo hello',
        extraLines: [
          '[Isolation] Environment: screen, Mode: attached',
          '[Isolation] Session: my-session',
        ],
      });

      expect(block).toContain('│ session   test-uuid');
      expect(block).toContain('│ isolation screen');
      expect(block).toContain('│ mode      attached');
      expect(block).toContain('│ screen    my-session');
      expect(block).toContain('$ echo hello');
    });

    it('should include docker metadata correctly', () => {
      const block = createStartBlock({
        sessionId: 'test-uuid',
        timestamp: '2025-01-01 00:00:00',
        command: 'echo hello',
        extraLines: [
          '[Isolation] Environment: docker, Mode: attached',
          '[Isolation] Image: ubuntu',
          '[Isolation] Session: docker-container-123',
        ],
      });

      expect(block).toContain('│ isolation docker');
      expect(block).toContain('│ image     ubuntu');
      expect(block).toContain('│ container docker-container-123');
    });
  });

  describe('createFinishBlock', () => {
    it('should create a finish block with result marker and metadata', () => {
      const block = createFinishBlock({
        sessionId: 'test-uuid-1234',
        timestamp: '2025-01-01 00:00:01',
        exitCode: 0,
        logPath: '/tmp/test.log',
        durationMs: 17,
      });

      expect(block).toContain('✓');
      expect(block).toContain('│ finish    2025-01-01 00:00:01');
      expect(block).toContain('│ duration  0.017s');
      expect(block).toContain('│ exit      0');
      expect(block).toContain('│ log       /tmp/test.log');
      expect(block).toContain('│ session   test-uuid-1234');
    });

    it('should use failure marker for non-zero exit code', () => {
      const block = createFinishBlock({
        sessionId: 'test-uuid',
        timestamp: '2025-01-01 00:00:01',
        exitCode: 1,
        logPath: '/tmp/test.log',
        durationMs: 100,
      });

      expect(block).toContain('✗');
      expect(block).toContain('│ exit      1');
    });

    it('should omit duration when not provided', () => {
      const block = createFinishBlock({
        sessionId: 'test-uuid-1234',
        timestamp: '2025-01-01 00:00:01',
        exitCode: 0,
        logPath: '/tmp/test.log',
      });

      expect(block).not.toContain('duration');
      expect(block).toContain('│ finish    2025-01-01 00:00:01');
    });

    it('should repeat isolation metadata in footer', () => {
      const block = createFinishBlock({
        sessionId: 'test-uuid',
        timestamp: '2025-01-01 00:00:01',
        exitCode: 0,
        logPath: '/tmp/test.log',
        durationMs: 17,
        extraLines: [
          '[Isolation] Environment: docker, Mode: attached',
          '[Isolation] Image: ubuntu',
          '[Isolation] Session: docker-container-123',
        ],
      });

      expect(block).toContain('│ isolation docker');
      expect(block).toContain('│ mode      attached');
      expect(block).toContain('│ image     ubuntu');
      expect(block).toContain('│ container docker-container-123');
    });

    it('should have log and session as the last two lines', () => {
      const block = createFinishBlock({
        sessionId: 'test-uuid',
        timestamp: '2025-01-01 00:00:01',
        exitCode: 0,
        logPath: '/tmp/test.log',
        durationMs: 17,
        extraLines: [
          '[Isolation] Environment: screen, Mode: attached',
          '[Isolation] Session: my-screen',
        ],
      });

      const lines = block.split('\n');
      expect(lines[lines.length - 1]).toBe('│ session   test-uuid');
      expect(lines[lines.length - 2]).toBe('│ log       /tmp/test.log');
    });
  });

  describe('formatDuration', () => {
    it('should format very small durations', () => {
      expect(formatDuration(0.5)).toBe('0.001s');
    });

    it('should format millisecond durations', () => {
      expect(formatDuration(17)).toBe('0.017s');
      expect(formatDuration(500)).toBe('0.500s');
    });

    it('should format second durations', () => {
      expect(formatDuration(1000)).toBe('1.000s');
      expect(formatDuration(5678)).toBe('5.678s');
    });

    it('should format longer durations with less precision', () => {
      expect(formatDuration(12345)).toBe('12.35s');
      expect(formatDuration(123456)).toBe('123.5s');
    });
  });

  describe('escapeForLinksNotation', () => {
    it('should not quote simple values', () => {
      expect(escapeForLinksNotation('simple')).toBe('simple');
      expect(escapeForLinksNotation('123')).toBe('123');
      expect(escapeForLinksNotation('true')).toBe('true');
    });

    it('should quote values with spaces', () => {
      expect(escapeForLinksNotation('hello world')).toBe('"hello world"');
    });

    it('should quote values with colons', () => {
      expect(escapeForLinksNotation('key:value')).toBe('"key:value"');
    });

    it('should use single quotes for values with double quotes', () => {
      expect(escapeForLinksNotation('say "hello"')).toBe('\'say "hello"\'');
    });

    it('should use double quotes for values with single quotes', () => {
      expect(escapeForLinksNotation("it's cool")).toBe('"it\'s cool"');
    });

    it('should escape quotes when both types are present', () => {
      const result = escapeForLinksNotation('say "hello" it\'s');
      // Should wrap in one quote type and escape the other
      expect(result).toMatch(/^["'].*["']$/);
    });

    it('should handle null values', () => {
      expect(escapeForLinksNotation(null)).toBe('null');
      expect(escapeForLinksNotation(undefined)).toBe('null');
    });
  });

  describe('formatAsNestedLinksNotation', () => {
    it('should format simple objects', () => {
      const obj = { key: 'value', number: 123 };
      const result = formatAsNestedLinksNotation(obj);

      expect(result).toContain('key value');
      expect(result).toContain('number 123');
    });

    it('should quote values with spaces', () => {
      const obj = { message: 'hello world' };
      const result = formatAsNestedLinksNotation(obj);

      expect(result).toContain('message "hello world"');
    });

    it('should handle empty objects', () => {
      expect(formatAsNestedLinksNotation({})).toBe('()');
    });

    it('should handle null', () => {
      expect(formatAsNestedLinksNotation(null)).toBe('null');
    });
  });
});

console.log('=== Output Blocks Unit Tests ===');

File: rust/tests/output_blocks_test.rs
//! Tests for output_blocks module
//!
//! Tests for the "timeline" format: width-independent, lossless output.

use start_command::{
    create_finish_block, create_start_block, escape_for_links_notation, format_duration,
    get_result_marker, parse_isolation_metadata, FinishBlockOptions, StartBlockOptions,
    FAILURE_MARKER, SUCCESS_MARKER, TIMELINE_MARKER,
};

#[test]
fn test_timeline_constants() {
    assert_eq!(TIMELINE_MARKER, "│");
    assert_eq!(SUCCESS_MARKER, "✓");
    assert_eq!(FAILURE_MARKER, "✗");
}

#[test]
fn test_get_result_marker() {
    assert_eq!(get_result_marker(0), "✓");
    assert_eq!(get_result_marker(1), "✗");
    assert_eq!(get_result_marker(127), "✗");
}

#[test]
fn test_parse_isolation_metadata() {
    let extra_lines = vec![
        "[Isolation] Environment: docker, Mode: attached",
        "[Isolation] Session: docker-container-123",
        "[Isolation] Image: ubuntu:latest",
    ];
    let metadata = parse_isolation_metadata(&extra_lines);

    assert_eq!(metadata.isolation, Some("docker".to_string()));
    assert_eq!(metadata.mode, Some("attached".to_string()));
    assert_eq!(metadata.session, Some("docker-container-123".to_string()));
    assert_eq!(metadata.image, Some("ubuntu:latest".to_string()));
}

#[test]
fn test_create_start_block() {
    let block = create_start_block(&StartBlockOptions {
        session_id: "test-uuid",
        timestamp: "2025-01-01 00:00:00",
        command: "echo hello",
        extra_lines: None,
        style: Some("rounded"),
        width: Some(50),
        defer_command: false,
    });

    assert!(block.contains("│ session   test-uuid"));
    assert!(block.contains("│ start     2025-01-01 00:00:00"));
    assert!(block.contains("$ echo hello"));
}

#[test]
fn test_create_start_block_with_isolation() {
    let extra = vec![
        "[Isolation] Environment: screen, Mode: attached",
        "[Isolation] Session: my-session",
    ];
    let block = create_start_block(&StartBlockOptions {
        session_id: "test-uuid",
        timestamp: "2025-01-01 00:00:00",
        command: "echo hello",
        extra_lines: Some(extra),
        style: Some("rounded"),
        width: Some(60),
        defer_command: false,
    });

    assert!(block.contains("│ session   test-uuid"));
    assert!(block.contains("│ start     2025-01-01 00:00:00"));
    assert!(block.contains("│ isolation screen"));
    assert!(block.contains("│ mode      attached"));
    assert!(block.contains("│ screen    my-session"));
    assert!(block.contains("$ echo hello"));
}

#[test]
fn test_create_start_block_with_docker_isolation() {
    let extra = vec![
        "[Isolation] Environment: docker, Mode: attached",
        "[Isolation] Image: ubuntu",
        "[Isolation] Session: docker-container-123",
    ];
    let block = create_start_block(&StartBlockOptions {
        session_id: "test-uuid",
        timestamp: "2025-01-01 00:00:00",
        command: "echo hello",
        extra_lines: Some(extra),
        style: None,
        width: None,
        defer_command: false,
    });

    assert!(block.contains("│ isolation docker"));
    assert!(block.contains("│ image     ubuntu"));
    assert!(block.contains("│ container docker-container-123"));
}

#[test]
fn test_create_start_block_with_defer_command() {
    // When defer_command is true, the command line should be omitted
    // (used when virtual commands will be shown separately)
    let extra = vec![
        "[Isolation] Environment: docker, Mode: attached",
        "[Isolation] Image: alpine:latest",
        "[Isolation] Session: docker-container-456",
    ];
    let block = create_start_block(&StartBlockOptions {
        session_id: "test-uuid",
        timestamp: "2025-01-01 00:00:00",
        command: "echo hello",
        extra_lines: Some(extra),
        style: None,
        width: None,
        defer_command: true,
    });

    assert!(block.contains("│ session   test-uuid"));
    assert!(block.contains("│ isolation docker"));
    assert!(block.contains("│ image     alpine:latest"));
    assert!(
        !block.contains("$ echo hello"),
        "Command should not appear when defer_command is true"
    );
}

#[test]
fn test_create_finish_block() {
    let block = create_finish_block(&FinishBlockOptions {
        session_id: "test-uuid",
        timestamp: "2025-01-01 00:00:01",
        exit_code: 0,
        log_path: "/tmp/test.log",
        duration_ms: Some(17.0),
        result_message: None,
        extra_lines: None,
        style: Some("rounded"),
        width: Some(60),
    });

    assert!(block.contains("✓"));
    assert!(block.contains("│ finish    2025-01-01 00:00:01"));
    assert!(block.contains("│ duration  0.017s"));
    assert!(block.contains("│ exit      0"));
    assert!(block.contains("│ log       /tmp/test.log"));
    assert!(block.contains("│ session   test-uuid"));
}

#[test]
fn test_create_finish_block_failure() {
    let block = create_finish_block(&FinishBlockOptions {
        session_id: "test-uuid",
        timestamp: "2025-01-01 00:00:01",
        exit_code: 1,
        log_path: "/tmp/test.log",
        duration_ms: Some(100.0),
        result_message: None,
        extra_lines: None,
        style: None,
        width: None,
    });

    assert!(block.contains("✗"));
    assert!(block.contains("│ exit      1"));
}

#[test]
fn test_create_finish_block_with_isolation_repeated() {
    let extra = vec![
        "[Isolation] Environment: docker, Mode: attached",
        "[Isolation] Image: ubuntu",
        "[Isolation] Session: docker-container-123",
    ];
    let block = create_finish_block(&FinishBlockOptions {
        session_id: "test-uuid",
        timestamp: "2025-01-01 00:00:01",
        exit_code: 0,
        log_path: "/tmp/test.log",
        duration_ms: Some(17.0),
        result_message: None,
        extra_lines: Some(extra),
        style: None,
        width: None,
    });

    assert!(block.contains("│ isolation docker"));
    assert!(block.contains("│ mode      attached"));
    assert!(block.contains("│ image     ubuntu"));
    assert!(block.contains("│ container docker-container-123"));
}

#[test]
fn test_create_finish_block_without_duration() {
    let block = create_finish_block(&FinishBlockOptions {
        session_id: "test-uuid",
        timestamp: "2025-01-01 00:00:01",
        exit_code: 0,
        log_path: "/tmp/test.log",
        duration_ms: None,
        result_message: None,
        extra_lines: None,
        style: Some("rounded"),
        width: Some(50),
    });

    assert!(block.contains("│ finish    2025-01-01 00:00:01"));
    assert!(!block.contains("duration"));
}

#[test]
fn test_finish_block_log_session_last() {
    let extra = vec![
        "[Isolation] Environment: screen, Mode: attached",
        "[Isolation] Session: my-screen",
    ];
    let block = create_finish_block(&FinishBlockOptions {
        session_id: "test-uuid",
        timestamp: "2025-01-01 00:00:01",
        exit_code: 0,
        log_path: "/tmp/test.log",
        duration_ms: Some(17.0),
        result_message: None,
        extra_lines: Some(extra),
        style: None,
        width: None,
    });

    let lines: Vec<&str> = block.lines().collect();
    assert_eq!(lines[lines.len() - 1], "│ session   test-uuid");
    assert_eq!(lines[lines.len() - 2], "│ log       /tmp/test.log");
}

#[test]
fn test_format_duration() {
    assert_eq!(format_duration(0.5), "0.001s");
    assert_eq!(format_duration(17.0), "0.017s");
    assert_eq!(format_duration(500.0), "0.500s");
    assert_eq!(format_duration(1000.0), "1.000s");
    assert_eq!(format_duration(5678.0), "5.678s");
    assert_eq!(format_duration(12345.0), "12.35s");
    assert_eq!(format_duration(123456.0), "123.5s");
}

#[test]
fn test_escape_for_links_notation_no_quoting() {
    assert_eq!(escape_for_links_notation("simple"), "simple");
    assert_eq!(escape_for_links_notation("123"), "123");
}

#[test]
fn test_escape_for_links_notation_with_space() {
    assert_eq!(escape_for_links_notation("hello world"), "\"hello world\"");
}

#[test]
fn test_escape_for_links_notation_with_colon() {
    assert_eq!(escape_for_links_notation("key:value"), "\"key:value\"");
}

#[test]
fn test_escape_for_links_notation_with_double_quotes() {
    assert_eq!(
        escape_for_links_notation("say \"hello\""),
        "'say \"hello\"'"
    );
}

#[test]
fn test_escape_for_links_notation_with_single_quotes() {
    assert_eq!(escape_for_links_notation("it's cool"), "\"it's cool\"");
}

// Issue #73: Visual Continuity Tests
#[test]
fn test_start_block_ends_with_empty_timeline_when_defer_command_true() {
    // Issue #73: When deferCommand is true (docker isolation),
    // the start block should end with │ (empty timeline line)
    // and no trailing empty line should be added by the CLI
    let extra = vec![
        "[Isolation] Environment: docker, Mode: attached",
        "[Isolation] Image: alpine:latest",
        "[Isolation] Session: docker-container-123",
    ];
    let block = create_start_block(&StartBlockOptions {
        session_id: "test-uuid",
        timestamp: "2026-01-08 12:00:00",
        command: "echo hello",
        extra_lines: Some(extra),
        style: None,
        width: None,
        defer_command: true,
    });

    let lines: Vec<&str> = block.lines().collect();
    // Last line should be empty timeline line (│)
    assert_eq!(
        lines[lines.len() - 1],
        "│",
        "Start block with defer_command should end with empty timeline line"
    );
    // Should not contain the command
    assert!(
        !block.contains("$ echo hello"),
        "Command should not appear when defer_command is true"
    );
}

#[test]
fn test_visual_continuity_start_block_for_docker_isolation() {
    // Issue #73: Visual continuity - the start block with docker isolation
    // should end cleanly for virtual commands to follow immediately
    let extra = vec![
        "[Isolation] Environment: docker, Mode: attached",
        "[Isolation] Image: alpine:latest",
        "[Isolation] Session: docker-1234",
    ];
    let block = create_start_block(&StartBlockOptions {
        session_id: "uuid-abc",
        timestamp: "2026-01-08 12:00:00",
        command: "echo hi",
        extra_lines: Some(extra),
        style: None,
        width: None,
        defer_command: true,
    });

    // Expected structure:
    // │ session   uuid-abc
    // │ start     2026-01-08 12:00:00
    // │
    // │ isolation docker
    // │ mode      attached
    // │ image     alpine:latest
    // │ container docker-1234
    // │   <-- ends here with empty timeline line

    let lines: Vec<&str> = block.lines().collect();
    assert_eq!(
        lines[lines.len() - 1],
        "│",
        "Should end with empty timeline line"
    );
    assert!(
        block.contains("│ container docker-1234"),
        "Should contain container name"
    );
    assert!(!block.contains("$ echo hi"), "Command should be deferred");
}

// Issue #73: Visual Continuity Tests
#[test]
fn test_visual_continuity_output_formatting() {
    // Issue #73: All commands should have consistent formatting:
    // 1. Command line ($ ...)
    // 2. Empty line (visual separation)
    // 3. Command output
    // 4. Empty line (visual separation)
    // 5. Result marker (✓ or ✗)
    //
    // This test documents the expected output structure.

    use start_command::{
        create_command_line, create_virtual_command_block, create_virtual_command_result,
    };

    // Verify createCommandLine produces the correct format
    let command_line = create_command_line("docker pull alpine:latest");
    assert_eq!(command_line, "$ docker pull alpine:latest");

    // Verify createVirtualCommandBlock matches createCommandLine
    let virtual_command_line = create_virtual_command_block("docker pull alpine:latest");
    assert_eq!(virtual_command_line, command_line);

    // Verify result markers
    assert_eq!(create_virtual_command_result(true), "✓");
    assert_eq!(create_virtual_command_result(false), "✗");
}

#[test]
fn test_visual_continuity_command_output_pattern() {
    // Issue #73: The expected output pattern for docker isolation with virtual commands:
    //
    // │ session   uuid-abc
    // │ start     2026-01-08 12:00:00
    // │
    // │ isolation docker
    // │ mode      attached
    // │ image     alpine:latest
    // │ container docker-1234
    // │
    // $ docker pull alpine:latest
    //
    // latest: Pulling from library/alpine
    // ...
    //
    // ✓
    // │
    // $ echo hi
    //
    // hi
    //
    // ✓
    //
    // The key aspects tested here:
    // 1. Start block ends with │ when defer_command is true
    // 2. Virtual command and user command follow the same pattern:
    //    command -> empty line -> output -> empty line -> result marker

    let extra = vec![
        "[Isolation] Environment: docker, Mode: attached",
        "[Isolation] Image: alpine:latest",
        "[Isolation] Session: docker-1234",
    ];
    let start_block = create_start_block(&StartBlockOptions {
        session_id: "uuid-abc",
        timestamp: "2026-01-08 12:00:00",
        command: "echo hi",
        extra_lines: Some(extra),
        style: None,
        width: None,
        defer_command: true,
    });

    // When defer_command is true, start block should end with empty timeline line
    let lines: Vec<&str> = start_block.lines().collect();
    assert_eq!(
        lines.last().unwrap(),
        &"│",
        "Start block with defer_command should end with │"
    );
}
